src/value.rs

```diff
@@ -0,0 +1,128 @@
+            waffle::Operator::I64Const { value } => Ok(WasmVal::I64(value as u64)),
+            waffle::Operator::F32Const { value } => Ok(WasmVal::F32(value)),
+            waffle::Operator::F64Const { value } => Ok(WasmVal::F64(value)),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum AbstractValue {
+    /// "top" default value; undefined.
+    #[default]
+    Top,
+    /// A value known at specialization time.
+    Concrete(WasmVal),
+    /// A value that points to memory known at specialization time,
+    /// with the given offset.
+    ConcreteMemory(MemoryBufferIndex, u32),
+    /// Static memory pointer.
+    StaticMemory(u32),
+    /// A value only computed at runtime. The instruction that
+    /// computed it is specified, if known.
+    Runtime(Option<waffle::Value>),
+}
+
+/// Memory pointed to by one of the incoming arguments to a
+/// specialized function.
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct MemoryBufferIndex(pub u32);
+
+impl AbstractValue {
+    pub(crate) fn meet(a: &AbstractValue, b: &AbstractValue) -> AbstractValue {
+        match (a, b) {
+            (AbstractValue::Top, x) | (x, AbstractValue::Top) => x.clone(),
+            (x, y) if x == y => x.clone(),
+            (AbstractValue::Concrete(a), AbstractValue::Concrete(b)) if a == b => {
+                AbstractValue::Concrete(*a)
+            }
+            (AbstractValue::Runtime(cause1), AbstractValue::Runtime(cause2)) => {
+                log::debug!(
+                    "runtime({:?} meet runtime({:?}) -> runtime({:?})",
+                    cause1,
+                    cause2,
+                    cause1.or(*cause2)
+                );
+                AbstractValue::Runtime(cause1.or(*cause2))
+            }
+            (AbstractValue::Runtime(cause1), _x) | (_x, AbstractValue::Runtime(cause1)) => {
+                AbstractValue::Runtime(*cause1)
+            }
+            (_av1, _av2) => AbstractValue::Runtime(None),
+        }
+    }
+
+    pub(crate) fn as_const_u32(&self) -> Option<u32> {
+        match self {
+            &AbstractValue::Concrete(WasmVal::I32(k)) => Some(k),
+            &AbstractValue::StaticMemory(addr) => Some(addr),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn as_const_u32_or_mem_offset(&self) -> Option<u32> {
+        match self {
+            &AbstractValue::Concrete(WasmVal::I32(k)) => Some(k),
+            &AbstractValue::ConcreteMemory(_, off) => Some(off),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn as_const_u64(&self) -> Option<u64> {
+        match self {
+            &AbstractValue::Concrete(WasmVal::I64(k)) => Some(k),
+            &AbstractValue::StaticMemory(addr) => Some(u64::from(addr)),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn as_const_truthy(&self) -> Option<bool> {
+        self.as_const_u32().map(|k| k != 0)
```

Comment:
Would it make sense to assert that the value is a `WasmVal::I32` here, if it's concrete? It seems like we shoudn't ever get a concrete value with a different type if we're expecting to evaluate it as a boolean.

Reply:
Overall I've tried to be ever so slightly panic-shy, erring toward "we don't know" / "runtime", as I'm a little scared of e.g. const-prop down dynamically unreachable paths or similar causing a panic of the whole toolchain rather than producing semantically still valid code (by returning `None` here, resulting in us not branch-folding). In other words I agree that *should* be the case, but I can't 100% convince myself it will be. Happy to still add it if you'd prefer though.

Reply:
Sounds good.